### PR TITLE
Add map operator

### PR DESCRIPTION
**Pull request by M0rtyMerr**

### CODE REVIEW

File: Sources/OpenCombine/Publishers/Publishers.Map.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
+//
+//  Publishers.Just.swift
```
Comment (broadwaylamb) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: `Publishers.Map.swift`

File: Sources/OpenCombine/Publishers/Publishers.Map.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+//  Created by Anton Nazarov on 25.06.2019.
+//
+
+extension Publisher {
+
+    /// Transforms all elements from the upstream publisher with a provided closure.
+    ///
+    /// - Parameter transform: A closure that takes one element as its parameter and returns a new element.
+    /// - Returns: A publisher that uses the provided closure to map elements from the upstream publisher to new elements that it then publishes.
+    public func map<T>(_ transform: @escaping (Self.Output) -> T) -> Publishers.Map<Self, T> {
```
Comment (broadwaylamb) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: Let's remove `Self` in `Self.Output` — it's redundant.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+final class MapTests: XCTestCase {
+
+    static let allTests = [
+        ("testJust", testJust),
+        ("testEmpty", testEmpty),
+        ("testError", testError),
+        ("testRange", testRange)
+    ]
+
+    func testJust() {
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift]: I suggest that we don't use `Just`, `Empty` and `Fail` publishers here, and use either `PassthroughSubject` or `CustomPublisher` type instead.

The reason is that `Just`, `Empty` and `Fail` have their own specializations for the `map` operator that have nothing to do with the `Map` struct, for instance:
https://github.com/broadwaylamb/OpenCombine/blob/17f9b0a29975106d2f51d5d9b7858c5b1e612019/Sources/OpenCombine/Publishers/Publishers.Just.swift#L203-L205

This means that these tests do not actually test the `Map` type. Those specializations are already tested here:
https://github.com/broadwaylamb/OpenCombine/blob/eab03def585705da0dc33b96763235a3669fa867/Tests/OpenCombineTests/PublisherTests/JustTests.swift#L304-L306

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        map.subscribe(tracking)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
+        XCTAssertEqual(tracking.history, [
+            .subscription(Subscriptions.empty),
+            .completion(Subscribers.Completion<TestingError>.failure(expectedError))
+        ])
+    }
+
+    func testRange() {
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift:57]: This is a good test!

If you have time, it'd be great to also add tests that ensure that demand is properly handled — something like this for `DropWhile`:
https://github.com/broadwaylamb/OpenCombine/blob/17f9b0a29975106d2f51d5d9b7858c5b1e612019/Tests/OpenCombineTests/PublisherTests/DropWhileTests.swift#L90-L129

Note that in the DropWhile case the request-demand logic was probably different from Map.

You may want to steal some more tests from the DropWhile suite :)

**Review by broadwaylamb — COMMENTED** [review]
Thanks @MortyMerr!

Some minor comments regarding tests.

Also please remove the `Map` declaration from the `RemainingCombineInterface.swift` file.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 3 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Sure, thanks, didn't noticed it

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 4 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift:57]: Sure. Added few more. Please review

**Comment by M0rtyMerr** [comment]
@broadwaylamb what do you think about the latest updates? :)
I see few people blocked with missing of `.map` so I am ready to apply any changes today

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    ]
+
+    func testEmpty() {
+        // Given
+        let tracking = TrackingSubscriberBase<Never>(receiveSubscription: { $0.request(.unlimited) })
+        let publisher = PassthroughSubject<Int, Never>()
+        // When
+        publisher.map { $0 * 2 }.subscribe(tracking)
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift]: This assert is redundant, as `tracking.subscriptions` is just a lazy view on `tracking.history`, which is already checked by the assert below.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    func testError() {
+        // Given
+        let expectedError = TestingError.oops
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.unlimited) })
+        let publisher = PassthroughSubject<Int, TestingError>()
+        // When
+        publisher.map { $0 * 2 }.subscribe(tracking)
+        publisher.send(completion: .failure(expectedError))
+        // Then
+        XCTAssertEqual(tracking.subscriptions.count, 1)
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Same here

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        ])
+    }
+
+    func testDemandSubscribe() {
+        // Given
+        let expectedSubscribeDemand = 42
+        let subscription = CustomSubscription()
+        let publisher = CustomPublisher(subscription: subscription)
+        let map = publisher.map { $0 * 2 }
+        let tracking = TrackingSubscriber(receiveSubscription: { $0.request(.max(expectedSubscribeDemand)) })
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Here you test that if a downstream subscriber (`tracking`) requests some elements in the `receiveSubscription` callback, the same number of elements is requested from the upstream publisher.

This is useful, but let's also make sure that if the downstream subscriber doesn't request any demand at all, the upstream publisher won't receive any requests.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            downstreamSubscription = $0
+        })
+        // When
+        map.subscribe(tracking)
+        try XCTUnwrap(downstreamSubscription).cancel()
+        // Then
+        XCTAssertEqual(subscription.history, [.requested(.unlimited), .canceled])
+    }
+
+    func testCancelAlreadyCancelled() throws {
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift:154]: This test's name is slightly misleading, I don't see that the subscription is cancelled twice. It makes sense to either update the name of the test (having in mind that `allTests` should also be updated), or add one more test for actual double-cancel.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        let publisher = CustomPublisher(subscription: subscription)
+        let map = publisher.map { $0 * 2 }
+        var downstreamSubscription: Subscription?
+        let tracking = TrackingSubscriber(receiveSubscription: {
+            $0.request(.unlimited)
+            downstreamSubscription = $0
+        })
+        // When
+        map.subscribe(tracking)
+        try XCTUnwrap(downstreamSubscription).cancel()
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/PublisherTests/MapTests.swift:166]: Good use of `XCTUnwrap`!

**Review by broadwaylamb — COMMENTED** [review]
I like the updates! More minor comments regarding tests incoming.

Note that I've added SwiftLint to the CI system, you may want to rebase onto master and fix the new warnings.

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 5 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Deleted

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 6 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Deleted

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 8 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift:154]: Added second Cancellation, that was the idea of test

File: Tests/OpenCombineTests/PublisherTests/MapTests.swift (hunk 7 above)
Comment (M0rtyMerr) [reply Tests/OpenCombineTests/PublisherTests/MapTests.swift]: Implemented additional test case

**Review by broadwaylamb — APPROVED** [review]

**Pull request merged**